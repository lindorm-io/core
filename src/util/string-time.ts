import { isFinite } from "lodash";

enum Duration {
  YEARS = "years",
  MONTHS = "months",
  DAYS = "days",
  HOURS = "hours",
  MINUTES = "minutes",
  SECONDS = "seconds",
  MILLISECONDS = "milliseconds",
}

type DurationObject = {
  [Duration.YEARS]: number;
  [Duration.MONTHS]: number;
  [Duration.DAYS]: number;
  [Duration.HOURS]: number;
  [Duration.MINUTES]: number;
  [Duration.SECONDS]: number;
  [Duration.MILLISECONDS]: number;
};

type RegExpObject = Record<Duration, RegExp>;

const regexp: RegExpObject = {
  [Duration.YEARS]: /(\d+ years)/g,
  [Duration.MONTHS]: /(\d+ months)/g,
  [Duration.DAYS]: /(\d+ days)/g,
  [Duration.HOURS]: /(\d+ hours)/g,
  [Duration.MINUTES]: /(\d+ minutes)/g,
  [Duration.SECONDS]: /(\d+ seconds)/g,
  [Duration.MILLISECONDS]: /(\d+ milliseconds)/g,
};

const getNumber = (string: string, regex: RegExp): number => {
  try {
    const result = string.toLowerCase().match(regex);

    if (!result || !result.length || result.length > 1) {
      return 0;
    }

    const number = parseInt(result[0].replace(/\s+/g, ""));

    if (!isFinite(number)) {
      return 0;
    }

    return number;
  } catch (_) {
    return 0;
  }
};

export const stringToDurationObject = (string: string): DurationObject => {
  const object: Record<Duration, number> = {
    [Duration.YEARS]: 0,
    [Duration.MONTHS]: 0,
    [Duration.DAYS]: 0,
    [Duration.HOURS]: 0,
    [Duration.MINUTES]: 0,
    [Duration.SECONDS]: 0,
    [Duration.MILLISECONDS]: 0,
  };

  for (const key of Object.keys(regexp)) {
    object[key as Duration] = getNumber(string, regexp[key as Duration]);
  }

  return object;
};

export const stringToMilliseconds = (string: string): number => {
  const object = stringToDurationObject(string);
  let time = 0;

  time = time + object[Duration.MILLISECONDS];
  time = time + object[Duration.SECONDS] * 1000;
  time = time + object[Duration.MINUTES] * 60 * 1000;
  time = time + object[Duration.HOURS] * 60 * 60 * 1000;
  time = time + object[Duration.DAYS] * 24 * 60 * 60 * 1000;
  time = time + object[Duration.MONTHS] * 30 * 24 * 60 * 60 * 1000;
  time = time + object[Duration.YEARS] * 365 * 24 * 60 * 60 * 1000;

  return time;
};

export const stringToSeconds = (string: string): number => {
  return Math.round(stringToMilliseconds(string) / 1000);
};
