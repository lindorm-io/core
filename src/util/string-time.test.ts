import { stringToDurationObject, stringToMilliseconds, stringToSeconds } from "./string-time";

describe("string-time.ts", () => {
  describe("stringToDurationObject", () => {
    test("should return years", () => {
      expect(stringToDurationObject("2 years").years).toBe(2);
    });

    test("should return months", () => {
      expect(stringToDurationObject("12 months").months).toBe(12);
    });

    test("should return days", () => {
      expect(stringToDurationObject("30 days").days).toBe(30);
    });

    test("should return hours", () => {
      expect(stringToDurationObject("24 hours").hours).toBe(24);
    });

    test("should return minutes", () => {
      expect(stringToDurationObject("60 minutes").minutes).toBe(60);
    });

    test("should return seconds", () => {
      expect(stringToDurationObject("15 seconds").seconds).toBe(15);
    });

    test("should return a full object", () => {
      expect(stringToDurationObject("1 years 2 months 3 days 4 hours 5 minutes 6 seconds")).toStrictEqual({
        years: 1,
        months: 2,
        days: 3,
        hours: 4,
        minutes: 5,
        seconds: 6,
      });
    });

    test("should return 0 on wrong or missing input", () => {
      expect(stringToDurationObject("100 milliseconds")).toStrictEqual({
        years: 0,
        months: 0,
        days: 0,
        hours: 0,
        minutes: 0,
        seconds: 0,
      });
    });
  });

  describe("stringToSeconds", () => {
    test("should return years", () => {
      expect(stringToSeconds("1 years")).toBe(31536000);
    });

    test("should return months", () => {
      expect(stringToSeconds("2 months")).toBe(5184000);
    });

    test("should return days", () => {
      expect(stringToSeconds("3 days")).toBe(259200);
    });

    test("should return hours", () => {
      expect(stringToSeconds("4 hours")).toBe(14400);
    });

    test("should return minutes", () => {
      expect(stringToSeconds("5 minutes")).toBe(300);
    });

    test("should return seconds", () => {
      expect(stringToSeconds("6 seconds")).toBe(6);
    });

    test("should return a combined number", () => {
      expect(stringToSeconds("1 years 2 months 3 days 4 hours 5 minutes 6 seconds")).toBe(36993906);
    });
  });

  describe("stringToMilliseconds", () => {
    test("should return years", () => {
      expect(stringToMilliseconds("1 years")).toBe(31536000000);
    });

    test("should return months", () => {
      expect(stringToMilliseconds("2 months")).toBe(5184000000);
    });

    test("should return days", () => {
      expect(stringToMilliseconds("3 days")).toBe(259200000);
    });

    test("should return hours", () => {
      expect(stringToMilliseconds("4 hours")).toBe(14400000);
    });

    test("should return minutes", () => {
      expect(stringToMilliseconds("5 minutes")).toBe(300000);
    });

    test("should return seconds", () => {
      expect(stringToMilliseconds("6 seconds")).toBe(6000);
    });

    test("should return a combined number", () => {
      expect(stringToMilliseconds("1 years 2 months 3 days 4 hours 5 minutes 6 seconds")).toBe(36993906000);
    });
  });
});
