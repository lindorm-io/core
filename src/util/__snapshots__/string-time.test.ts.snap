// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string-time.ts stringToDurationObject should return 0 on wrong or missing input 1`] = `
Object {
  "days": 0,
  "hours": 0,
  "minutes": 0,
  "months": 0,
  "seconds": 0,
  "years": 0,
}
`;

exports[`string-time.ts stringToDurationObject should return a full object 1`] = `
Object {
  "days": 3,
  "hours": 4,
  "minutes": 5,
  "months": 2,
  "seconds": 6,
  "years": 1,
}
`;
